content/performance/preallocate-arrays.md

```diff
@@ -0,0 +1,41 @@
+---
+title: "Preallocate Arrays"
+section: "Performance"
+menu:
+  toc:
+    parent: "performance"
+    weight: 20
+---
+
+## Problem
+
+Your code is performance sensitive and suffers from poor performance due to arrays being resized while in use.
+
+## Solution
+
+It's common for new Pony programmers to create arrays that use the default allocation:
+
+```pony
+// allocate space for no entries
+let small: Array[U8] = Array[U8]
+```
+
+Instead, preallocate more space than you expect to use:
+
+```pony
+// preallocate space for at least 2048 entries
+let not_so_small: Array[U8] = Array[U8](2048)
+```
+
+## Discussion
+
+Arrays are an incredibly flexible data structure. Programmers can deploy them in situations where the amount of data needed will vary over time. If the array is too small and more space is needed to hold additional items, the array will be expanded so that it can hold more items.
+
+This expansion of an array is often called "reallocating." In general, you want to avoid reallocating because:
+
+- a new chunk of memory is allocated
+- the contents of the old array are copied over into the newly allocated space
```

Comment:
This makes it sound like the old contents are always copied. Is this the case? I thought realloc first tries to enlarge the memory allocation pointed to by the original pointer, in which case it wouldn't need to copy anything. But perhaps I've misunderstood.

Reply:
I'm not sure that level of detail is warranted. It's why I pretty much wanted to send them off to look at the source. I'm open for suggestions on how to improve without it becoming a morass.

Reply:
I'd suggest something like "the contents of the old array might be copied over into the newly allocated space". The way it is currently written could mislead people into thinking the copying always takes place. I don't think you need to go into more detail than that.